pact/dao/dao.pact

```diff
@@ -0,0 +1,279 @@
+(enforce-pact-version "3.7")
+
+(namespace (read-msg 'dao-ns))
+
+(module utils-v1 GOVERNANCE
+  (defcap GOVERNANCE () false)
+  (defun now:time () (at "block-time" (chain-data)))
+
+  (defun today:time ()
+    (parse-time "%Y-%m-%d" (format-time "%Y-%m-%d" (now))))
+
+  (defun yesterday:time () (add-time (today) (days (- 1))))
+
+  (defun tomorrow:time () (add-time (today) (days 1)))
+
+  (defun btwn-incl:bool (start end ts)
+    (enforce (>= end start) (format "bounding start {} must come before end {}" [start end]))
+    (and (>= ts start) (<= ts end)))
+  (defun btwn-excl:bool (start end ts)
+    (enforce (>= end start) "bounding end comes before start")
+    (and (> ts start) (< ts end)))
+
+  (defun this-block:integer () (at 'block-height (chain-data)))
+
+  (defconst BLOCKHEIGHT_PER_MINUTE 2)
+  (defconst BLOCKHEIGHT_PER_HOUR 120)
+  (defconst BLOCKHEIGHT_PER_DAY 2880)
+
+  (defun row-with-key (tbl k)
```

Comment:
Pact's typesystem can't handle this yet. Suggest one of two approaches:
- just put the key in the rowdata if it's that important. It doesn't look like there will be a lot of data so don't over-optimize.
- if you really need it, specialize for the specific table. While valid pact, it isn't really set up for passing tables around.

Normally the best practice is to avoid table scans/`select` etc by managing an index in the form of a list column somewhere. However it seems like your dataset is small so I would just add keys to the schemas. It makes life easier anyway.

Reply:
I like that approach -- just add the key to the object.